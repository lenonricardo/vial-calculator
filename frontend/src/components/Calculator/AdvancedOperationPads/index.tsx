import { TopPadWrapper, TopPad } from './styles'

interface OperationPadsProps {
  onInput: (input: string) => void
}

enum AdvancedOperationEnum {
  ALL_CLEAR = 'AC',
  SQUARE_ROOT = '√',
  EXPONENTIATION = 'xʸ',
  PERCENTAGE = '%'
}

export default function AdvancedOperationPads({ onInput }: OperationPadsProps) {
  const advancedOperationPads: AdvancedOperationEnum[] = Object.values(
    AdvancedOperationEnum
  )

  const handleOperationClick = (operation: string) => {
    onInput(operation)
  }

  return (
    <TopPadWrapper>
      {advancedOperationPads.map((label) => (
        <TopPad key={label} onClick={() => handleOperationClick(label)}>
          {label}
        </TopPad>
      ))}
    </TopPadWrapper>
  )
}
