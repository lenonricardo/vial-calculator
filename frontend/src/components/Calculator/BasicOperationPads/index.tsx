import {
  Wrapper,
  PrimaryPad,
  PrimaryGradientPad,
  SecondaryGradientPad
} from './styles'

interface OperationPadsProps {
  onInput: (input: string) => void
  onEquals: () => void
}

enum BasicOperationsEnum {
  DIVISION = '/',
  MULTIPLICATION = '*',
  SUBSTRACTION = '-',
  ADDITION = '+',
  EQUALS = '='
}

export default function OperationPads({
  onInput,
  onEquals
}: OperationPadsProps) {
  const handleOperationClick = (operation: string) => {
    onInput(operation)
  }

  const handleEquals = () => {
    onEquals()
  }

  return (
    <Wrapper>
      <PrimaryPad
        onClick={() => handleOperationClick(BasicOperationsEnum.DIVISION)}
      >
        {BasicOperationsEnum.DIVISION}
      </PrimaryPad>

      <PrimaryGradientPad
        onClick={() => handleOperationClick(BasicOperationsEnum.MULTIPLICATION)}
      >
        {BasicOperationsEnum.MULTIPLICATION}
      </PrimaryGradientPad>

      <SecondaryGradientPad
        onClick={() => handleOperationClick(BasicOperationsEnum.SUBSTRACTION)}
      >
        {BasicOperationsEnum.SUBSTRACTION}
      </SecondaryGradientPad>

      <SecondaryGradientPad
        onClick={() => handleOperationClick(BasicOperationsEnum.ADDITION)}
      >
        {BasicOperationsEnum.ADDITION}
      </SecondaryGradientPad>

      <SecondaryGradientPad onClick={() => handleEquals()}>
        {BasicOperationsEnum.EQUALS}
      </SecondaryGradientPad>
    </Wrapper>
  )
}
