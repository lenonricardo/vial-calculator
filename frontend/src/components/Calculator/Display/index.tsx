import {
  Wrapper,
  Header,
  CustomHistoryIcon,
  Result,
  CustomBackspaceIcon
} from './styles'

type DisplayProps = {
  result: string
}

export default function Display(props: DisplayProps) {
  return (
    <Wrapper>
      <Header>
        <CustomHistoryIcon></CustomHistoryIcon>
      </Header>

      <Result>{props.result || 0}</Result>
      <CustomBackspaceIcon />
    </Wrapper>
  )
}
