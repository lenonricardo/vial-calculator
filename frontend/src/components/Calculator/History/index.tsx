import {
  CustomDeleteForeverIcon,
  Expression,
  Total,
  Wrapper,
} from './styles'
import React, { useState, useEffect } from 'react'


interface History {
  expression: string,
  total: string,
}

export default function History() {

  const [history, setHistory] = useState<Array<History>>([])

  useEffect(() => {
    const operationHistory = getHistory()

    setHistory(operationHistory?.slice(0, 4) || [])
  }, [])

  const getHistory = () => {
    const storedHistory: string | null = localStorage.getItem('history')

    if (storedHistory) {
      return JSON.parse(storedHistory)
    }
  }

  const clearHistory = () => {
    localStorage.removeItem('history')
    setHistory([])
  }

  return (
    <Wrapper>
      <div>
        {history.map((result, i) => (
          <>
            <Expression key={i}>{result.expression} =</Expression>
            <Total key={i}>{result.total}</Total>
          </>
        ))}
      </div>

      {
        !!history.length &&
        <CustomDeleteForeverIcon onClick={clearHistory}/>
      }
    </Wrapper>
  )
}
