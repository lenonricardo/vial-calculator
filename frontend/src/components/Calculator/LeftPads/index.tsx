import { PadWrapper, TopPad, Pads, BottomPads, DoublePad, Pad } from './styles'

enum TopPadEnum {
  CE = 'CE',
  C = 'C',
  CHANGE_SIGN = '+/-'
}

enum NumberPadEnum {
  SEVEN = '7',
  EIGHT = '8',
  NINE = '9',
  FOUR = '4',
  FIVE = '5',
  SIX = '6',
  ONE = '1',
  TWO = '2',
  THREE = '3'
}

enum BottomPadEnum {
  ZERO = '0',
  DECIMAL_POINT = '.'
}

export default function LeftPads() {
  const topPads: TopPadEnum[] = Object.values(TopPadEnum)
  const numberPads: NumberPadEnum[] = Object.values(NumberPadEnum)

  return (
    <PadWrapper>
      <Pads>
        {topPads.map((label) => (
          <TopPad key={label}>{label}</TopPad>
        ))}

        {numberPads.map((label) => (
          <Pad key={label}>{label}</Pad>
        ))}
      </Pads>

      <BottomPads>
        <DoublePad>{BottomPadEnum.ZERO}</DoublePad>
        <Pad>{BottomPadEnum.DECIMAL_POINT}</Pad>
      </BottomPads>
    </PadWrapper>
  )
}
