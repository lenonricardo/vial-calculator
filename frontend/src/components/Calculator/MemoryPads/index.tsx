import { MemoryWrapper, SecondaryPad } from './styles'

interface MemoryPadsProps {
  onInput: (input: string) => void
}

enum MemoryPadsEnum {
  MEMORY_PLUS = 'M+',
  MEMORY_MINUS = 'M-',
  MEMORY_RECALL = 'MR',
  MEMORY_CLEAR = 'MC'
}

export default function MemoryPads({ onInput }: MemoryPadsProps) {
  const memoryPads: MemoryPadsEnum[] = Object.values(MemoryPadsEnum)

  const handleOperationClick = (operation: string) => {
    onInput(operation)
  }

  return (
    <MemoryWrapper>
      {memoryPads.map((label) => (
        <SecondaryPad key={label}>{label}</SecondaryPad>
      ))}
    </MemoryWrapper>
  )
}
