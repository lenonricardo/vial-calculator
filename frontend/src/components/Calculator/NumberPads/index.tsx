import { PadWrapper, Pads, BottomPads, DoublePad, Pad } from './styles'

interface NumberPadsProps {
  onInput: (input: string) => void
}

enum NumberPadEnum {
  SEVEN = '7',
  EIGHT = '8',
  NINE = '9',
  FOUR = '4',
  FIVE = '5',
  SIX = '6',
  ONE = '1',
  TWO = '2',
  THREE = '3'
}

enum BottomPadEnum {
  ZERO = '0',
  DECIMAL_POINT = '.'
}

export default function NumberPads({ onInput }: NumberPadsProps) {
  const numberPads: NumberPadEnum[] = Object.values(NumberPadEnum)

  const handleNumberClick = (number: string) => {
    onInput(number)
  }

  return (
    <PadWrapper>
      <Pads>
        {numberPads.map((label) => (
          <Pad key={label} onClick={() => handleNumberClick(label)}>
            {label}
          </Pad>
        ))}
      </Pads>

      <BottomPads>
        <DoublePad onClick={() => handleNumberClick(BottomPadEnum.ZERO)}>
          {BottomPadEnum.ZERO}
        </DoublePad>
        <Pad onClick={() => handleNumberClick(BottomPadEnum.DECIMAL_POINT)}>
          {BottomPadEnum.DECIMAL_POINT}
        </Pad>
      </BottomPads>
    </PadWrapper>
  )
}
