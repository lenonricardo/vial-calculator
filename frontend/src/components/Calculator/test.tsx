import { render } from '@testing-library/react'

import Calculator from '.'

describe('<Calculator />', () => {
  it('should render the component', () => {
    const { container } = render(<Calculator />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
