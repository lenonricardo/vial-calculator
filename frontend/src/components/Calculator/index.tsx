import React, { useState } from 'react'
import Display from './Display'
import NumberPads from './NumberPads'
import BasicOperationPads from './BasicOperationPads'
import { Wrapper, PadWrapper } from './styles'
import { evaluate } from 'mathjs'
import MemoryPads from './MemoryPads'
import AdvancedOperationPads from './AdvancedOperationPads'

const Calculator = () => {
  const [expression, setExpression] = useState<string>('')
  const [result, setResult] = useState<string>('')

  const handleInput = (input: string) => {
    setResult((prevExpression) => prevExpression + input)

    // if (result.includes('+')) {
    //   const evaluatedResult = evaluate(result);
    //   setExpression(evaluatedResult)
    // }
  }

  const handleOperationInput = (input: string) => {
    setResult((prevExpression) => prevExpression + input)

    // const evaluatedResult = evaluate(result);
    // setExpression(evaluatedResult)
  }

  const handleClear = () => {
    setExpression('')
    setResult('')
  }

  const handleCancelEntry = () => {
    setResult((prevExpression) => prevExpression.slice(0, -1))
  }

  const handleEquals = () => {
    try {
      const evaluatedResult = evaluate(result)
      setResult(evaluatedResult.toString())
      setExpression('')
    } catch (error) {
      setResult('Error')
    }
  }

  return (
    <Wrapper>
      <Display result={result} />

      <AdvancedOperationPads onInput={handleInput} />

      <PadWrapper>
        <NumberPads onInput={handleInput} />
        <MemoryPads onInput={handleOperationInput} />
        <BasicOperationPads
          onInput={handleOperationInput}
          onEquals={handleEquals}
        />
      </PadWrapper>
    </Wrapper>
  )
}

export default Calculator
