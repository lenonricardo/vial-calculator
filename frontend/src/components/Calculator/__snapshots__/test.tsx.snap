// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator /> should render the component 1`] = `
.c1 {
  width: var(--calculator-width);
  height: var(--calculator-height);
  border-radius: 1.5rem 1.5rem 0 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  font-size: var(--default-font-size);
  margin-right: 10px;
  color: var(--theme-primary);
  cursor: pointer;
}

.c4 {
  color: var(--font-result);
  font-size: var(--result-font-size);
  margin-right: 10px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  padding: 10px;
}

.c3 {
  font-size: var(--default-font-size);
  margin-right: 10px;
  color: var(--theme-primary);
  cursor: pointer;
}

.c7 {
  width: var(--pad-width);
  height: var(--pad-height);
  color: var(--font-pads);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: var(--default-font-size);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c11 {
  -webkit-transform: translateY(var(--pad-width));
  -ms-transform: translateY(var(--pad-width));
  transform: translateY(var(--pad-width));
  display: grid;
  grid-template-columns: repeat(auto-fill,minmax(var(--pad-width),1fr));
  grid-template-rows: repeat(auto-fill,minmax(var(--pad-width),1fr));
  width: calc(450px - var(--pad-width) - var(--pad-width));
  max-height: calc(450px - var(--pad-width));
}

.c13 {
  width: calc(var(--pad-width) * 2);
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c17 {
  background: var(--theme-primary);
  color: #fff !important;
}

.c18 {
  background: linear-gradient( 180deg, var(--theme-primary) 9%, var(--theme-secondary) 100% );
  color: #fff !important;
}

.c19 {
  background: var(--theme-secondary);
  color: #fff !important;
}

.c19:last-child {
  border-radius: 0 0 1.5rem 0;
}

.c0 {
  width: var(--calculator-width);
  background-color: #fff;
  border-radius: 1.5rem;
  box-shadow: rgba(0,0,0,0.35) 0px 5px 15px;
}

.c9 {
  width: var(--calculator-width);
  border-radius: 0 0 1.5rem 1.5rem;
  display: -webkit-box;
  display: -webkit-f;
  display: -ms-fbox;
  display: f;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-transform: translateY(var(--pad-width));
  -ms-transform: translateY(var(--pad-width));
  transform: translateY(var(--pad-width));
}

.c15 {
  background-color: var(--secondary-operations);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-transform: translateY(var(--pad-width));
  -ms-transform: translateY(var(--pad-width));
  transform: translateY(var(--pad-width));
  margin-top: calc(-1 * var(--pad-width));
}

.c8 {
  background-color: var(--secondary-operations);
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium c3 css-i4bv87-MuiSvgIcon-root"
        data-testid="HistoryIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 3c-4.97 0-9 4.03-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42C8.27 19.99 10.51 21 13 21c4.97 0 9-4.03 9-9s-4.03-9-9-9zm-1 5v5l4.28 2.54.72-1.21-3.5-2.08V8H12z"
        />
      </svg>
    </div>
    <div
      class="c4"
    >
      0
    </div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium c5 css-i4bv87-MuiSvgIcon-root"
      data-testid="BackspaceIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M22 3H7c-.69 0-1.23.35-1.59.88L0 12l5.41 8.11c.36.53.9.89 1.59.89h15c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-3 12.59L17.59 17 14 13.41 10.41 17 9 15.59 12.59 12 9 8.41 10.41 7 14 10.59 17.59 7 19 8.41 15.41 12 19 15.59z"
      />
    </svg>
  </div>
  <div
    class="c6"
  >
    <div
      class="c7 c8"
    >
      AC
    </div>
    <div
      class="c7 c8"
    >
      √
    </div>
    <div
      class="c7 c8"
    >
      xʸ
    </div>
    <div
      class="c7 c8"
    >
      %
    </div>
  </div>
  <div
    class="c9"
  >
    <div
      class="c10"
    >
      <div
        class="c11"
      >
        <div
          class="c7"
        >
          7
        </div>
        <div
          class="c7"
        >
          8
        </div>
        <div
          class="c7"
        >
          9
        </div>
        <div
          class="c7"
        >
          4
        </div>
        <div
          class="c7"
        >
          5
        </div>
        <div
          class="c7"
        >
          6
        </div>
        <div
          class="c7"
        >
          1
        </div>
        <div
          class="c7"
        >
          2
        </div>
        <div
          class="c7"
        >
          3
        </div>
      </div>
      <div
        class="c12"
      >
        <div
          class="c7 c13"
        >
          0
        </div>
        <div
          class="c7"
        >
          .
        </div>
      </div>
    </div>
    <div
      class="c14"
    >
      <div
        class="c7 c15"
      >
        M+
      </div>
      <div
        class="c7 c15"
      >
        M-
      </div>
      <div
        class="c7 c15"
      >
        MR
      </div>
      <div
        class="c7 c15"
      >
        MC
      </div>
    </div>
    <div
      class="c16"
    >
      <div
        class="c7 c17"
      >
        /
      </div>
      <div
        class="c7 c18"
      >
        *
      </div>
      <div
        class="c7 c19"
      >
        -
      </div>
      <div
        class="c7 c19"
      >
        +
      </div>
      <div
        class="c7 c19"
      >
        =
      </div>
    </div>
  </div>
</div>
`;
